lib/molinillo/resolution.rb

```diff
@@ -213,36 +220,211 @@ def initial_state
+            unwind_details[:index] = candidate_index
+            unwind_details[:relationship] = :primary
+            unwind_details[:requirement] = r
+            unwind_details[:conflicting_requirements] = binding_requirements
+            return unwind_details if unwind_details[:index] == maximal_index
+
+            next # No need to look at this requirement's parent, as it couldn't have a higher index
+          end
+
+          # Next, look at the parent of this requirement, and check if the requirement
+          # could have been avoided if an alternative PossibilitySet had been chosen
+          parent_r = parent_of(r)
+          requirement_state = find_state_for(parent_r)
+          candidate_index = states.index(requirement_state)
+          next unless candidate_index && candidate_index >= unwind_details[:index]
+          if requirement_state &&
+              requirement_state.possibilities.any? { |set| set.dependencies.none? { |dep| dep == r } }
+            if candidate_index > unwind_details[:index] || unwind_details[:relationship] == :primary
+              unwind_details[:index] = candidate_index
+              unwind_details[:relationship] = :parent
+              unwind_details[:requirement] = r
+              unwind_details[:conflicting_requirements] = binding_requirements
+            end
+
+            next # No need to look at this requirement's grandparent, as it couldn't have a higher index
+          end
+
+          # Finally, look at the grandparent and up of this requirement, looking
+          # for any possibilities at all - we're now too far removed from the conflict
+          # to know how any such possibility will affect it, so just have to try it
+          grandparent_r = parent_of(parent_r)
+          until grandparent_r.nil?
+            requirement_state = find_state_for(grandparent_r)
+            candidate_index = states.index(requirement_state)
+            break unless candidate_index && candidate_index >= unwind_details[:index]
+            if requirement_state && !requirement_state.possibilities.empty?
+              unwind_details[:index] = candidate_index
+              unwind_details[:relationship] = :grandparent
+              unwind_details[:requirement] = r
+              unwind_details[:conflicting_requirements] = binding_requirements
               break
             end
-            r = parent_of(r)
+            grandparent_r = parent_of(grandparent_r)
           end
         end
 
-        index
+        unwind_details
+      end
+
+      # @param [DependencyState] state
+      # @param [Array] array of requirements
+      # @return [Boolean] whether or not the given state has any possibilities
+      #    that could satisfy the given requirements
+      def conflict_fixing_possibilities?(state, binding_requirements)
+        return false unless state
+
+        state.possibilities.map(&:possibilities).flatten(1).any? do |poss|
```

Comment:
since we don't use this array, my guess is that it might be a slight performance win to use nested `any?` statements rather than having to `map` and `flatten`